File: make-test-data/go/main_test.go
package main

import (
	"encoding/csv"
	"os"
	"strconv"
	"strings"
	"testing"
	"time"
)

func TestGenerateCustomers(t *testing.T) {
	numCustomers := 10
	customers := generateCustomers(numCustomers)

	if len(customers) != numCustomers {
		t.Errorf("Expected %d customers, got %d", numCustomers, len(customers))
	}

	for i, customer := range customers {
		if customer.ID != i+1 {
			t.Errorf("Expected customer ID %d, got %d", i+1, customer.ID)
		}
		if len(customer.Name) == 0 {
			t.Errorf("Expected customer name to be non-empty")
		}
		if len(customer.Email) == 0 {
			t.Errorf("Expected customer email to be non-empty")
		}
	}
}

func TestGenerateOrders(t *testing.T) {
	numCustomers := 10
	minOrders := 5
	maxOrders := 10
	customers := generateCustomers(numCustomers)
	statusList := []string{"Pending", "Processing", "Shipped", "Delivered"}

	orders := generateOrders(customers, minOrders, maxOrders, statusList, true)
	if len(orders) < numCustomers*minOrders || len(orders) > numCustomers*maxOrders {
		t.Errorf("Expected number of orders between %d and %d, got %d", numCustomers*minOrders, numCustomers*maxOrders, len(orders))
	}

	statuses := map[string]bool{"Pending": true, "Processing": true, "Shipped": true, "Delivered": true}
	for i, order := range orders {
		if order.ID != i+1 {
			t.Errorf("Expected order ID %d, got %d", i+1, order.ID)
		}
		if order.CustomerID < 1 || order.CustomerID > numCustomers {
			t.Errorf("Order customer ID %d out of range", order.CustomerID)
		}
		if len(order.ProductName) == 0 {
			t.Errorf("Expected product name to be non-empty")
		}
		if order.Quantity < 1 || order.Quantity > 9 {
			t.Errorf("Order quantity %d out of range", order.Quantity)
		}
		if _, err := time.Parse("2006-01-02", order.OrderDate); err != nil {
			t.Errorf("Order date %s is not a valid date", order.OrderDate)
		}
		if !statuses[order.Status] {
			t.Errorf("Invalid order status %s", order.Status)
		}
	}

	orders = generateOrders(customers, minOrders, maxOrders, statusList, false)
	for i, order := range orders {
		if order.Status != statusList[i%len(statusList)] {
			t.Errorf("Expected order status %s, got %s", statusList[i%len(statusList)], order.Status)
		}
	}
}

func TestWriteCSV(t *testing.T) {
	data := [][]string{
		{"ID", "Name"},
		{"1", "Alice"},
		{"2", "Bob"},
	}

	filename := "test.csv"
	err := writeCSV(filename, data)
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}
	defer os.Remove(filename)

	file, err := os.Open(filename)
	if err != nil {
		t.Fatalf("Expected no error opening file, got %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		t.Fatalf("Expected no error reading file, got %v", err)
	}

	if len(records) != len(data) {
		t.Errorf("Expected %d records, got %d", len(data), len(records))
	}

	for i, record := range records {
		for j, field := range record {
			if field != data[i][j] {
				t.Errorf("Expected %s, got %s", data[i][j], field)
			}
		}
	}
}

File: make-test-data/rust/tests/main_test.rs
use std::fs::File;
use std::io::{self, Read};
use tempfile::tempdir;

use data_generation::{generate_customers, generate_orders, write_csv};

#[test]
fn test_generate_customers() {
    let num_customers = 10;
    let customers = generate_customers(num_customers);
    assert_eq!(customers.len(), num_customers);
    for (i, customer) in customers.iter().enumerate() {
        assert_eq!(customer.id, i + 1);
        assert!(!customer.name.is_empty());
        assert!(customer.email.ends_with("@example.com"));
    }
}

#[test]
fn test_generate_orders() {
    let num_customers = 10;
    let min_orders = 5;
    let max_orders = 10;
    let customers = generate_customers(num_customers);
    let statuses = vec!["Pending", "Processing", "Shipped", "Delivered"];

    let orders = generate_orders(&customers, min_orders, max_orders, &statuses, true);
    assert!(orders.len() >= num_customers * min_orders);
    assert!(orders.len() <= num_customers * max_orders);

    let valid_statuses: Vec<String> = statuses.into_iter().map(|s| s.to_string()).collect();
    for (i, order) in orders.iter().enumerate() {
        assert_eq!(order.id, i + 1);
        assert!(order.customer_id >= 1 && order.customer_id <= num_customers);
        assert!(!order.product_name.is_empty());
        assert!(order.quantity >= 1 && order.quantity <= 10);
        assert!(valid_statuses.contains(&order.status));
    }

    let orders = generate_orders(&customers, min_orders, max_orders, &valid_statuses, false);
    for (i, order) in orders.iter().enumerate() {
        assert_eq!(order.status, valid_statuses[i % valid_statuses.len()]);
    }
}

#[test]
fn test_write_csv() -> io::Result<()> {
    let data = vec![
        vec!["ID".to_string(), "Name".to_string()],
        vec!["1".to_string(), "Alice".to_string()],
        vec!["2".to_string(), "Bob".to_string()],
    ];
    let dir = tempdir()?;
    let file_path = dir.path().join("test.csv");
    write_csv(&file_path, &data)?;

    let mut file = File::open(file_path)?;
    let mut contents = String::new();
    file.read_to_string(&mut contents)?;
    let expected = "ID,Name\n1,Alice\n2,Bob\n";
    assert_eq!(contents, expected);

    dir.close()?;
    Ok(())
}